src/libxrpl/protocol/UintTypes.cpp

```diff
@@ -132,4 +132,44 @@ badCurrency()
     return currency;
 }
 
+bool
+validCurrency(Currency const& currency, PaymentTx paymentTx)
+{
+    // Allow payments for invalid non-standard currencies
+    // created pre fixNonStandardCurrency amendment
+    static std::set<Currency> whiteList = {
+        Currency{"0000000000000000000000000000000078415059"},
+        Currency{"00000000004150756E6B30310000000000000000"},
+        Currency{"0000000000D9A928EFBCBEE297A1EFBCBE29DBB6"},
+        Currency{"0000000000414C6F676F30330000000000000000"},
+        Currency{"0000000000000000000000005852500000000000"},
+        Currency{"000028E0B2A05FE0B2A029E2948CE288A9E29490"},
+        Currency{"00000028E2989EEFBE9FE28880EFBE9F29E2989E"},
+        Currency{"00000028E381A3E29795E280BFE2979529E381A3"},
+        Currency{"000000000000005C6D2F5F283E5F3C295F5C6D2F"},
+        Currency{"00000028E295AFC2B0E296A1C2B0EFBC89E295AF"},
+        Currency{"0000000000000000000000005852527570656500"},
+        Currency{"000000000000000000000000302E310000000000"},
+        Currency{"0000000000E1839A28E0B2A05FE0B2A0E1839A29"},
+        Currency{"0000000048617070794E6577596561725852504C"},
+        Currency{"0000E29D9AE29688E29590E29590E29688E29D9A"},
+        Currency{"000028E297A35FE297A229E2948CE288A9E29490"},
+        Currency{"00000000CA95E0B2A0E0B2BFE1B4A5E0B2A0CA94"},
+        Currency{"000000282D5F282D5F282D5F2D295F2D295F2D29"},
+        Currency{"0000000000000000000000000000000078415049"},
+        Currency{"00000000000028E295ADE0B2B05FE280A2CC8129"}};
```

Comment:
An interesting thing might happen if someone creates a new invalid currency during the voting period of this amendment. Even worse if it happens during the two-week period after voting has completed. I do not think there's anything we can do about it, other than allow errors to happen on transactions with such an invalid currency in the future - I think there should be **no** future PR to extend this list.

Reply:
Yes, I agree. There should be no future PR to extend the list. But it might make sense to check if there are more invalid currencies created before this PR is merged. This list is good as of couple of weeks ago.

Reply:
It would be trivial to create thousands of additional entries for this list until the PR is merged. The way currency codes are being interpreted is up to client applications and should not be part of the protocol.

See also the documentation: "To prevent this from **being treated as** a "standard" currency code" - there is nothing stating that "standard currency codes" are a rule.

If you want to enforce standards, you could also start blocking non-ISO-4217 currency codes. This would include BTC, ETH and XRP by the way.

Reply:
There are about 140,000 currencies codes out of which ~4,700 are standard and ~135,300 are non-standard. The 22 non-standard currencies that don't follow the rule were clearly created unintentionally. Creating thousands more  doesn't serve any purpose and is a violation of the rule.

Reply:
This is a guideline for interpreting a 160 bit field when using some APIs, not a rule. I also don't see any reason or benefit of limiting this in the protocol (especially with such a clunky allowlist approach, considering that there are more than one ledger out there using this software).

Reply:
What if, instead of a hard-coded whitelist, you put the whitelist on the ledger in a new singleton object that just holds an array of Currencies, much like the Amendments singleton?

It could use the `LedgerFix` transaction introduced in #4945 with a new `LedgerFixType`. Add fields to `LedgerFix` to specify the currency code, and one account that has an existing trust line (thus ensuring that this is a currency that already exists on the ledger, and nobody can whitelist a currency preemptively). Use the default base fee, since this isn't doing a lot of work. You've already got an amendment gating all of this, so you wouldn't need another one. It might be worth reusing the ideas from `Rules` so that the object doesn't have to keep being loaded, but this doesn't have to do the extra work that `Rules` does to handle presets, so the existing object caches may be enough - it isn't likely to expire from the cache since so many transactions will read it.

This would
1. Get rid of the hard-coded list.
2. Allow the whitelist to work across all networks.
3. Allow any new invalid currencies created before the amendment is enabled to also be whitelisted.
4. Keeping it cheap, should encourage issuers to pay the few drops to fix their currencies, but if they're not, anyone can do it. Whitelisting all the current ones would only cost 200 drops total.